File: crates/darkpool-matcher/src/change_note.rs
//! Role tags for the six note roles produced by a match.
//!
//! # What lives here now
//!
//! Only the role BYTES. They are consumed by VALID_MATCH_BATCH v3's Poseidon
//! derivations — user outputs are `Poseidon3(24, consumed_input_inner, role)`
//! and fee outputs are `Poseidon4(36, epoch_key, consumed_use_tag, role)` — so
//! the tags themselves are still very much live protocol constants.
//!
//! # What was removed (audit 2026-07-25, S-06)
//!
//! This module used to also export `derive_inner(match_id, role)`, a SHA-256
//! construction (`SHA256("darknyx-change-inner-v2" || match_id_le || role)`,
//! Fr-masked) that was the v2 way of deriving a note's `inner_hash`. CS-03
//! replaced it: outputs now derive from the CONSUMED INPUT rather than from a
//! match identifier, which is what removed caller-selected output randomness.
//!
//! The helper survived that migration with a full cross-language KAT still
//! green, which is precisely the failure shape the 2026-07-14 pass warned
//! about — a parity suite validating a retired construction against itself
//! while the live path used something else. `run_batch` was still calling it to
//! fill `note_e_commitment`/`note_f_commitment` with values the chain would
//! never create; the settle assembler overwrote them, but only when its own
//! Poseidon derivation succeeded, so an error there let the stale SHA value
//! through. It was also exported from the SDK's public index, where a consumer
//! following it would compute a commitment absent from the tree and see their
//! balance silently under-report.
//!
//! Change-note inners are now derived exclusively by
//! `darkpool_crypto::match_output::match_output_inner_hash`.

/// Role tag for the buyer's change note (`note_e`).
pub const CHANGE_ROLE_BUYER: u8 = 0xB1;
/// Role tag for the seller's change note (`note_f`).
pub const CHANGE_ROLE_SELLER: u8 = 0x5E;

// ─── Trade-output + fee note roles ───────────────────────────────────────────

/// Role tag for the buyer's full-fill output note (`note_c`).
pub const TRADE_ROLE_BUYER: u8 = 0xC1;
/// Role tag for the seller's full-fill output note (`note_d`).
pub const TRADE_ROLE_SELLER: u8 = 0xD1;
/// Role tag for a base-asset protocol fee note.
pub const FEE_ROLE_BASE: u8 = 0xFB;
/// Role tag for a quote-asset protocol fee note.
pub const FEE_ROLE_QUOTE: u8 = 0xFC;

File: packages/sdk/src/utxo/change-note.ts
/**
 * Change/trade note `inner_hash` derivation — SDK port of the matcher's
 * `darkpool_matcher::change_note::derive_inner` (and the on-chain
 * `tee_forced_settle`/assembler use of it).
 *
 * This SHA-based helper pins the legacy pure-matcher's provisional commitment
 * construction. The live VALID_MATCH_BATCH v3 settlement instead uses
 * `utxo/match-output.ts::deriveMatchOutputInner`, derived from the consumed
 * input inner, for every final and continuing output.
 *
 *   inner_hash = SHA-256("nyx-change-inner" ‖ match_id_u64_le ‖ role) , Fr-safe
 *                masked (byte 0 = 0, byte 1 high-nibble cleared).
 *
 * Byte-identical to the matcher; pinned by the cross-language KAT in
 * `tests/change-note-inner-parity.test.ts` (and the SDK test alongside it).
 */

import { createHash } from "node:crypto";

/** Role tags — must match `darkpool_matcher::change_note` + on-chain constants. */
export const CHANGE_ROLE_BUYER = 0xb1; // note_e (buyer's quote-side change)
export const CHANGE_ROLE_SELLER = 0x5e; // note_f (seller's base-side change)

/**
 * Derive the `inner_hash` (32-byte BE, Fr-safe) for a final change note of the
 * given `matchId` + `role`.
 */
export function deriveChangeInner(matchId: bigint, role: number): Uint8Array {
  if (matchId < 0n || matchId > 0xffff_ffff_ffff_ffffn) {
    throw new Error(`matchId must be a u64; got ${matchId}`);
  }
  if (!Number.isInteger(role) || role < 0 || role > 0xff) {
    throw new Error(`role must be a u8; got ${role}`);
  }
  const h = createHash("sha256");
  h.update(Buffer.from("nyx-change-inner"));
  const mid = new Uint8Array(8);
  new DataView(mid.buffer).setBigUint64(0, matchId, true); // little-endian
  h.update(mid);
  h.update(new Uint8Array([role]));
  const d = new Uint8Array(h.digest());
  d[0] = 0; // Fr-safe: clear the top byte …
  d[1] &= 0x0f; // … and the high nibble of byte 1.
  return d;
}
